dali/kernels/reduce/reduce_all_gpu_impl.cuh

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef _DALI_KERNELS_REDUCE_REDUCE_ALL_GPU_IMPL_CUH
+#define _DALI_KERNELS_REDUCE_REDUCE_ALL_GPU_IMPL_CUH
+
+
+#include "dali/core/util.h"
+#include "dali/kernels/reduce/reduce.h"
+
+namespace dali {
+namespace kernels {
+
+template <typename Acc, typename Reduction, typename OutputIdxFn>
+__device__ void BlockReduce(Acc *out, Acc val, Reduction reduce, OutputIdxFn get_output_idx) {
```

Comment:
Do you really need `get_output_idx` to be a function? Why a single value won't do?

Reply:
This way the warps that don't need the value (all with threadIdx.y > 0) won't even bother calculating it.